addons/stock/models/stock_quant.py

```diff
@@ -378,6 +470,24 @@ def _onchange_serial_number(self):
             if message:
                 return {'warning': {'title': _('Warning'), 'message': message}}
 
+    def _apply_inventory(self):
+        moves = OrderedSet()
+        for quant in self:
+            # Create and validate a move so that the quant matches its `inventory_quantity`.
+            if float_compare(quant.inventory_diff_quantity, 0, precision_rounding=quant.product_uom_id.rounding) > 0:
+                move_vals = quant._get_inventory_move_values(quant.inventory_diff_quantity, quant.product_id.with_company(
+                    quant.company_id).property_stock_inventory, quant.location_id)
+            else:
+                move_vals = quant._get_inventory_move_values(-quant.inventory_diff_quantity, quant.location_id, quant.product_id.with_company(
+                    quant.company_id).property_stock_inventory, out=True)
+            moves.add(quant.env['stock.move'].with_context(inventory_mode=False).create(move_vals).id)
+        self.env['stock.move'].browse(moves)._action_done()
+        self.location_id.write({'last_inventory_date': fields.Date.today()})
+        locations = self.mapped('location_id')
+        for location in locations:
+            self.filtered(lambda q: q.location_id == location).write({'inventory_quantity': 0, 'inventory_date': location._get_next_inventory_date()})
```

Comment:
to avoid possible performance bottleneck with filtered in a for loop.
```python3
date_by_location = {loc: loc._get_next_inventory_date() for loc in self.mapped('location_id')}
for quant in self:
    quant.inventory_date = date_by_location[quant.location_id]
self.write({'inventory_quantity': 0, 'inventory_diff_quantity': 0})
     
```

Reply:
Removed filter.

As for your date idea, your solution is much more elegant but I realized that we need to check if `next_inventory_date` <= `today` so it's not possible to do it your way because we need to know the max number of days in a month in order to handle if the user puts in a number greater than the max days (e.g. 100) and because of the leap year we need to check the max days for both `today.year` and the next year. I.e. leap years are annoying.